### DISCUSSION TITLE
hdu.setChecksum() error. User error?

### DISCUSSION THREAD

**Discussion by johnmurphyastro** [nom-tam-fits/nom-tam-fits#663] [General]
Hi all,
I am trying to read in a FITS image, modify the header and then write it to a new, different, file.
If I create a checksum, the checksum fails when the new image is read back.
If I don't create a checksum, no errors occur.
It still fails even if I don't modify any header entries.
Is this user error or a problem in the new release? 

```
File inputFile = new File("S:\\test\\inputImage.fit");
File outputFile = new File("S:\\test\\outputImage.fit");

try (Fits fits = new Fits(inputFile)) {
    BasicHDU<?> hdu = fits.getHDU(0);
    Header header = hdu.getHeader();
    // Add header edits here...
    //
    hdu.setChecksum();  // fits.setChecksum(); also fails
    fits.write(outputFile);
} catch (IOException | FitsException ex) {
    System.out.println(ex.getMessage());
}

// Try to read outputImage.fit
try (Fits fits = new Fits(outputFile)) {
    BasicHDU<?>[] HDUs = fits.read();
    for (BasicHDU<?> hdUs : HDUs) {
        hdUs.verifyIntegrity();
    }
} catch (IOException | FitsException ex) {
    System.out.println(ex.getMessage());
}
```
While reading outputImage.fit, a FitsExcepton is thrown and the following text was displayed:
Failed checksum: got 896016252, expected 4294967295

Is this error occurring because I have not copied all the data into new HDUs, and am using the same Fits object for both reading and writing the image?

Regards, John Murphy

**Comment by attipaci** [comment]
Hi John!

I'll look into it... In the meantime, can you see what happens if you add `hdu.getData().getKernel();` to before `Header header = hdu.getHeader();`?

**Reply by johnmurphyastro** [reply]
Hi @attipaci ,
I've just tried it. Unfortunately it makes no difference. It produces:
Failed checksum: got 778838644, expected 4294967295

There is something really strange though: 
1. The expected checksum is always the same
2. The got checksum is always the same while the program is running, but changes to a new value after restarting the program. Strange! So for example, after restarting the program:

Failed checksum: got 778838904, expected 4294967295

Thanks for looking into this, John Murphy

**Comment by johnmurphyastro** [comment]
For example, run1; read the same file three times:
Failed checksum: got 778838904, expected 4294967295
Failed checksum: got 778838904, expected 4294967295
Failed checksum: got 778838904, expected 4294967295
Restart the program, still read that same file three times:
Failed checksum: got 778838908, expected 4294967295
Failed checksum: got 778838908, expected 4294967295
Failed checksum: got 778838908, expected 4294967295
Restart again:
Failed checksum: got 778706301, expected 4294967295
Failed checksum: got 778706301, expected 4294967295
Failed checksum: got 778706301, expected 4294967295

**Comment by attipaci** [comment]
Note to anyone reading this discussion thread. It continues as Issue #664 and a solution was prepared in PR #665, which has been merged to the `master` branch and will be included in the upcoming 1.20.1 release in the coming weeks.
